File: simulator/trees/Proof.py

def verify_branch(root, index, proof, output_as_int=False):
# index = get_index_in_permuted(index, 2**len(proof) // 2)
index += 2**len(proof)
v = proof[0]
for p in proof[1:]:
    if index % 2:
        v = blake(p + v)
    else:
        v = blake(v + p)
    index //= 2
assert v == root
return int.from_bytes(proof[0], 'big') if output_as_int else proof[0]

# Make a compressed proof for multiple indices
def mk_multi_branch(tree, indices):
# Branches we are outputting
output = []
# Elements in the tree we can get from the branches themselves
calculable_indices = {}
for i in indices:
    new_branch = mk_branch(tree, i)
    index = len(tree) // 2 + i
    calculable_indices[index] = True
    for j in range(1, len(new_branch)):
        calculable_indices[index ^ 1] = True
        index //= 2
    output.append(new_branch)
# Fill in the calculable list: if we can get or calculate both children, we can calculate the parent
complete = False
while not complete:
    complete = True
    keys = sorted([x//2 for x in calculable_indices])[::-1]
    for k in keys:
        if k*2 in calculable_indices and (k*2)+1 in calculable_indices and k not in calculable_indices:
            calculable_indices[k] = True
            complete = False
# If for any branch node both children are calculable, or the node overlaps with a leaf, or the node
# overlaps with a previously calculated one, elide it
scanned = {}
for i, b in zip(indices, output):
    index = len(tree) // 2 + i
    scanned[index] = True
    for j in range(1, len(b)):
        if ((index^1)*2 in calculable_indices and (index^1)*2+1 in calculable_indices) or ((index^1)-len(tree)//2 in indices) or (index^1) in scanned:
            b[j] = b''
        scanned[index^1] = True
        index //= 2
return output

File: simulator/trees/Proof.js
function mkBranch(tree, index) {
  let i = index + Math.floor(tree.length / 2);
  let o = [tree[index]];
  while (i > 1) {
    o.push(tree[index ^ 1]);
    i = Math.floor(i / 2);
  }
  return o;
}

const tree = "0123456789abcdef";

const branch = mkBranch(tree, 0);

// console.log(branch);

function mkMultiBranch(tree, indices) {
  let output = [];
  let calculableIndices = {};
  indices.forEach((i) => {
    let newBranch = mkBranch(tree, i);
    let index = Math.floor(tree.length / 2) + i;
    calculableIndices[index.toString()] = true;
    for (let j = 1; j < newBranch.length; j++) {
      calculableIndices[(index ^ 1).toString()] = true;
      index = Math.floor(index / 2);
    }
    output.push(newBranch);
  });

  let complete = false;
  while (!complete) {
    complete = true;
    let keys = Object.keys(calculableIndices)
      .map((x) => {
        return Math.floor(x / 2);
      })
      .sort()
      .reverse();
    for (let k in keys) {
      if (
        Object.keys(calculableIndices).includes((k * 2).toString()) &
        Object.keys(calculableIndices).includes((k * 2 + 1).toString()) &
        !Object.keys(calculableIndices).includes(k.toString())
      ) {
        calculableIndices[k.toString()] = true;
        complete = false;
      }
    }
  }
  let scanned = {};
  let iandb = indices.map((index, idx) => {
    return {
      i: index,
      b: output[idx],
    };
  });
  iandb.forEach((i) => {
    let index = Math.floor(tree.length / 2) + i.i;
    scanned[index] = true;
    for (let j = 1; j < i.b.length; j++) {
      if (
        (Object.keys(calculableIndices).includes(((index ^ 1) * 2).toString()) &
          Object.keys(calculableIndices).includes(
            ((index ^ 1) * 2 + 1).toString()
          )) |
        indices.includes((index ^ 1) - Math.floor(tree.lenth / 2)) |
        Object.keys(scanned).includes((index ^ 1).toString())
      ) {
        i.b[j] = "0x";
      }
      scanned[index ^ 1] = true;
      index = Math.floor(index / 2);
    }
  });
  return output;
}

const multi = mkMultiBranch(tree, [15, 0]);

// console.log(multi);

// for (let i = 16; i < 24; i++) {
//   console.log(i ^ 1);
// }

const treeNodes = {
  16: { active: false },
  17: { active: false },
  18: { active: false },
  19: { active: false },
  20: { active: true },
  21: { active: false },
  22: { active: true },
  23: { active: false },
};

export default function multiProof(tree) {
  let proof = [];
  for (let i = 23; i >= 16; i--) {
    // iterate through leaf ndoes
    let neighboridx = i ^ 1;
    // and leaf's neighbor
    let node = tree[i.toString()];
    let neighbor = tree[neighboridx.toString()];
    if (node.active & !neighbor.active) {
      proof.push(neighboridx - 8);
    }
    if (node.active & neighbor.active) {
      proof.push(node-8);
      proof.push(neighboridx-8)
    }
  }
  proof.forEach((n) =>
    proof.push(
      Math.floor(n / 2) ^ 1,
      Math.floor((Math.floor(n / 2) ^ 1) / 2) ^ 1
    )
  );
  for (let i = 0; i < 15; i++) {
    if (proof.includes(i) & proof.includes(i ^ 1)) {
      proof = proof.filter((n) => (n != i));
      proof = proof.filter((n) => (n != (i^1)))
    }
  }

  return proof;
}

console.log(multiProof(treeNodes));

